### PR TITLE
generate_bear_requirements.py: Add language file

### PR DESCRIPTION
**Pull request by sladyn98**
This generates a yaml file mapping the languages used by every bear.

Closes https://github.com/coala/coala-bears/issues/2479

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: .ci/generate_bear_requirements.py [hunk 1]
```diff
@@ -163,6 +164,20 @@ def get_cabal_requirements(requirements):
     return _get_requirements(requirements, '==')
 
 
+def get_languages(bears):
+    language_dict={}
+<<<<<<< HEAD
```
Comment (shikharvaish28) [inline .ci/generate_bear_requirements.py]: You need to solve this merge conflict

Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: Sould i squash all commits

Comment (shikharvaish28) [reply .ci/generate_bear_requirements.py]: First of all, try to solve the conflict and then squash it. (In case you're not able to find where to solve the issue, try to squash them all, it will raise a merge conflict error which you can solve from your editor)

File: .ci/generate_bear_requirements.py [hunk 2]
```diff
@@ -163,6 +164,13 @@ def get_cabal_requirements(requirements):
     return _get_requirements(requirements, '==')
 
 
+def get_languages(bears):
+    language_dict={}
+    for bear in bears:
+        language_dict[str(bear.name)]=bear.LANGUAGES
```
Comment (li-boxuan) [inline .ci/generate_bear_requirements.py]: redundant `str()` ?

File: .ci/generate_bear_requirements.py [hunk 3]
```diff
@@ -258,6 +266,11 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    language_requirements = get_languages(get_all_bears(bear_dirs))
+    file_path = os.path.join(PROJECT_DIR,BEAR_LANGUAGE_REQUIREMENTS_YAML)
+    with open(file_path,'w') as outfile:
+        yaml.dump(language_requirements,outfile)
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply .ci/generate_bear_requirements.py]: Better sort them before dumping

File: .ci/generate_bear_requirements.py (hunk 2 above)
Comment (li-boxuan) [reply .ci/generate_bear_requirements.py]: I think it returns a string literal. Correct me if I am wrong.

Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: When i coded the solution and tested  it gave me an error so i put it there.

File: .ci/generate_bear_requirements.py [hunk 4]
```diff
@@ -258,6 +266,11 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    language_requirements = get_languages(get_all_bears(bear_dirs))
```
[the remark this answers is not in the document]
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: I need the language function to create a dictionary to dump in yaml.

**Comment by Dhiraj240** [comment]
@sladyn98 Its a basic requirement , you can adjust things like ``` Generate language file ```  as ```Add file or Add language```

File: bear-requirements.yaml [hunk 5]
```diff
@@ -172,10 +172,6 @@ pip_requirements:
     version: ~=3.4.0
   lxml:
     version: ==3.6.0
-  memento_client:
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply bear-requirements.yaml]: But it looks like those libraries are being used. Could you post the command how you get these removals?

Even if the removal is necessary, IMO you should create an issue instead of solving it in this pr. At least you should split it into another commit, since this is unrelated to your current commit.

File: .ci/generate_bear_requirements.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply .ci/generate_bear_requirements.py]: I mean you can reuse the result of this `get_all_bears(bear_dirs)`. See comment above.

File: bear-language-requirements.yaml [hunk 6]
```diff
@@ -0,0 +1,355 @@
+ApertiumLintBear: !!set
```
Comment (jayvdb) [inline bear-language-requirements.yaml]: we do not need these to be sets.

simple lists would be ok.

the set is coming from `bear.LANGUAGES`.  You should convert that to a list and sort it.

Comment (sladyn98) [reply bear-language-requirements.yaml]: I made the requested changes @jayvdb

File: .ci/generate_bear_requirements.py [hunk 7]
```diff
@@ -87,6 +88,7 @@ def get_all_bears(bear_dirs):
     return list(itertools.chain(local_bears, global_bears))
 
 
+
```
Comment (Naveenaidu) [inline .ci/generate_bear_requirements.py]: This extra line is unnecessary

File: .ci/generate_bear_requirements.py [hunk 8]
```diff
@@ -163,6 +165,13 @@ def get_cabal_requirements(requirements):
     return _get_requirements(requirements, '==')
 
 
+def get_languages(bears):
+    language_dict = {}
+    for bear in bears:
+        language_dict[str(bear.name)]=list(bear.LANGUAGES).sort()
+    return language_dict
+
+
```
Comment (Naveenaidu) [inline .ci/generate_bear_requirements.py]: This is unnecessary too.

File: .ci/generate_bear_requirements.py (hunk 7 above)
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: made the changes

File: bear-requirements.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply bear-requirements.yaml]: This is not addressed.

**Comment by sladyn98** [comment]
@li-boxuan i made the requested changes by adding those files back.

File: .ci/generate_bear_requirements.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply .ci/generate_bear_requirements.py]: This is not addressed.

File: bear-language-requirements.yaml (hunk 6 above)
Comment (li-boxuan) [reply bear-language-requirements.yaml]: This is not addressed.

File: .ci/generate_bear_requirements.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: @li-boxuan how do i squash them i cant access them from my terminal

[the remark this answers is not in the document]
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: I got it thanks i learnt a lot from this. It was because i kept pushing

[the remark this answers is not in the document]
Comment (ayan-b) [reply .ci/generate_bear_requirements.py]: Glad it helped. :blush:

File: bear-language-requirements.yaml [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+  - java
+  - objective-c
+  - php
+  - python
+  - python3
+  - tcl
+  - vhdl
+ElmLintBear:
+  - Elm
+AnnotationBear: []
```
Comment (li-boxuan) [inline bear-language-requirements.yaml]: Please also sort bears according to their names.

File: .ci/generate_bear_requirements.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply .ci/generate_bear_requirements.py]: Again, this is not addressed. If you disagree with me, you can comment here. Otherwise please make changes.

**Comment by ishanSrt** [comment]
Now I notice, many of the things that I pointed out have already been addressed by @li-boxuan but you didn't make the changes. Please address all your review before asking for more. 👍

File: .ci/generate_bear_requirements.py [hunk 10]
```diff
@@ -258,6 +268,12 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    language_requirements = get_languages(get_all_bears(bear_dirs))
+    file_path = os.path.join(PROJECT_DIR,BEAR_LANGUAGE_REQUIREMENTS_YAML)
```
Comment (ishanSrt) [inline .ci/generate_bear_requirements.py]: space after comma

File: tests/swift/TailorBearTest.py [hunk 11]
```diff
@@ -24,22 +24,22 @@
 }
 """
 
-TailorBearWithoutConfig = verify_local_bear(TailorBear,
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply tests/swift/TailorBearTest.py]: Seems these changes are from 7bc134e59873137fb1b9f0feacea3bce77d2ed80. Please use git properly.

File: .ci/generate_bear_requirements.py (hunk 10 above)
Comment (ishanSrt) [reply .ci/generate_bear_requirements.py]: is Gitmate bot taking a nap? 😅

File: .ci/generate_bear_requirements.py [hunk 12]
```diff
@@ -26,6 +26,8 @@
 from coalib.bears.BEAR_KIND import BEAR_KIND
 from coalib.collecting.Collectors import collect_bears
 
+from pyprint.ConsolePrinter import ConsolePrinter
```
Comment (ishanSrt) [inline .ci/generate_bear_requirements.py]: unused import

File: bear-language-requirements.yaml [hunk 13]
```diff
@@ -0,0 +1,411 @@
+ApertiumLintBear:
```
Comment (ishanSrt) [inline bear-language-requirements.yaml]: Please run YAMLLintBear on this file, since YAML is guided by indentation, I see inconsistent styles being used for the bears with no dependencies i.e. with `[]` in comparison to the rest of the file, which can be avoided easily

File: tests/swift/TailorBearTest.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sladyn98) [reply tests/swift/TailorBearTest.py]: @li-bouxuan can you kindly explain how do I get rid of this issue.

File: .travis.yml [hunk 14]
```diff
@@ -4,7 +4,6 @@ language: python
 python:
   - 3.4
   - 3.5
-  - 3.6
```
[the remark this answers is not in the document]
Comment (sladyn98) [reply .travis.yml]: This change was not done by me. Should i change it back

File: .ci/generate_bear_requirements.py [hunk 15]
```diff
@@ -162,6 +166,12 @@ def get_pip_requirements(requirements):
 def get_cabal_requirements(requirements):
     return _get_requirements(requirements, '==')
 
+def get_languages(bears):
+    language_dict = {}
+    for bear in bears:
+        language_dict[str(bear.name)]=list(sorted(bear.LANGUAGES))
```
[the remark this answers is not in the document]
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: Done

File: .ci/generate_bear_requirements.py [hunk 16]
```diff
@@ -258,6 +268,12 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    language_requirements = get_languages(get_all_bears(bear_dirs))
```
[the remark this answers is not in the document]
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: Yes I solved this

File: .ci/generate_bear_requirements.py (hunk 10 above)
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: Yes i made these changes.

File: .ci/generate_bear_requirements.py (hunk 12 above)
Comment (sladyn98) [reply .ci/generate_bear_requirements.py]: I removed this import

File: .ci/generate_bear_requirements.py [hunk 17]
```diff
@@ -258,6 +268,12 @@ def sort_requirements(req_dict):
...
+    file_path = os.path.join(PROJECT_DIR,BEAR_LANGUAGE_REQUIREMENTS_YAML)
+    with open(file_path, 'w') as outfile:
+        yaml.indent(mapping = 2, sequence = 4, offset = 2)
```
Comment (anctartica) [inline .ci/generate_bear_requirements.py]: is this required? the previous code doesn't need to use this.

File: .travis.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (anctartica) [reply .travis.yml]: yes please

**Comment by sladyn98** [comment]
@anctartica yes it is because i need proper indentation.

**Comment by sladyn98** [comment]
@li-boxuan  @ishanSrt  I made all the changes that you guys requested. Except i cant figure out about the test lint bear, because those changes have already been merged to master, they cant be removed.
Apart from that if any changes are required kindly suggest.

File: tests/swift/TailorBearTest.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply tests/swift/TailorBearTest.py]: You encountered this issue because you didn't rebase your commit properly. To solve this, revert your changes on `TailorBearTest.py` and `TSLintBearTest.py`.

**Comment by jayvdb** [comment]
Needs to be rebased properly, and commit message *must* follow https://coala.io/commit .  These are basics that *must* be done, otherwise CI fails, and it is not worth reviewer time to investigate problems with your PR because it is very probable that the problem is that you are using old code , not rebased.

**Comment by sladyn98** [comment]
@jayvdb I did make the requested chages. I rebased, I changed the commit message, I made the changes requested by @li-boxuan  and @ishanSrt thank you for teaching me these valuable things. The gitmate bot is giving me an error on some files but i guess that is a bug.

File: .coafile [hunk 18]
```diff
@@ -82,4 +82,4 @@ shellcheck_ignore = SC2046, SC2154, SC2155
 bears = Jinja2Bear
 files = **.jj2
 # Bug in handling of ".. list | sort"
-ignore = .moban.dt/Gemfile.jj2, .moban.dt/bear-requirements.txt.jj2, .moban.dt/package.json.jj2
+ignore = .moban.dt/Gemfile.jj2, .moban.dt/bear-requirements.txt.jj2, .moban.dt/package.json.jj2
```
Comment (li-boxuan) [inline .coafile]: Please remove unrelated change.

**Comment by sladyn98** [comment]
@li-boxuan dont you think that is a bug because in the original  coala repo the file is the same.

**Comment by li-boxuan** [comment]
The issue requires you to generate a mapping of languages to bears, but you are generating a mapping of bears to languages. You need to either explain on the issue page, or you need to follow the requirement.

**Comment by sladyn98** [comment]
@li-bouxan i spoke to John about this and he said it's fine.

**Comment by sladyn98** [comment]
@li-boxuan I addressed the chaneges the boxes are no longer there

**Comment by sladyn98** [comment]
@li-bouxan I made the changes.

File: bear-language-requirements.yaml (hunk 13 above)
Comment (ishanSrt) [reply bear-language-requirements.yaml]: The changes have been made by @sladyn98. Marking as resolved.

**Comment by jayvdb** [comment]
Rebase your patch before submitting it for review.

**Comment by jayvdb** [comment]
Please look at the Travis CI failures.  They show an obvious problem with your patch.

File: bear-language-requirements.yaml [hunk 19]
```diff
@@ -0,0 +1,411 @@
...
+DartLintBear:
+  - Dart
+DennisBear:
+  - po
+  - pot
+DocGrammarBear:
+  - c
+  - cpp
+  - cs
+  - default
```
Comment (jayvdb) [inline bear-language-requirements.yaml]: Interesting ... `default` is not a language.  Maybe this is a bug in the bear, or this PR needs to create a special case to skip that one.

File: bear-language-requirements.yaml [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+  - python3
+  - tcl
+  - vhdl
+DockerfileLintBear:
+  - Dockerfile
+DocumentationStyleBear:
+  - c
+  - cpp
+  - cs
+  - default
```
Comment (jayvdb) [inline bear-language-requirements.yaml]: again, 'default'

File: bear-language-requirements.yaml [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+  - JSX
+  - JavaScript
+JavaPMDBear:
+  - Java
+Jinja2Bear:
+  - Jinja2
+JuliaLintBear:
+  - Julia
+KeywordBear:
+  - All
```
Comment (jayvdb) [inline bear-language-requirements.yaml]: 'All' isnt a language.  It should be omitted from the yaml file.  This should be a special case.

maybe write these as

```yaml
KeywordBear: true
```

to indicate the bear exists, but has no defined languages.
`KeywordBear: false` is also a possible way of describing these bears.

**Comment by sladyn98** [comment]
> Please look at the Travis CI failures. They show an obvious problem with your patch.
git diff is exiting with code 1

File: bear-language-requirements.yaml [hunk 22]
```diff
@@ -0,0 +1,411 @@
...
+CSharpLintBear:
+  - C#
+CheckstyleBear:
+  - Java
+ClangASTPrintBear:
+  - C
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
Comment (DanPi32) [inline bear-language-requirements.yaml]:
```suggestion
  - Objective-CPP
```

File: bear-language-requirements.yaml [hunk 23]
```diff
@@ -0,0 +1,411 @@
...
+  - OpenCL
+  - OpenMP
+ClangBear:
+  - C
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
Comment (DanPi32) [inline bear-language-requirements.yaml]:
```suggestion
  - Objective-CPP
```

File: bear-language-requirements.yaml [hunk 24]
```diff
@@ -0,0 +1,411 @@
...
+  - OpenCL
+  - OpenMP
+ClangCloneDetectionBear:
+  - C
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
Comment (DanPi32) [inline bear-language-requirements.yaml]:
```suggestion
  - Objective-CPP
```

File: bear-language-requirements.yaml [hunk 25]
```diff
@@ -0,0 +1,411 @@
...
+  - OpenCL
+  - OpenMP
+ClangComplexityBear:
+  - C
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
Comment (DanPi32) [inline bear-language-requirements.yaml]:
```suggestion
  - Objective-CPP
```

File: bear-language-requirements.yaml [hunk 26]
```diff
@@ -0,0 +1,411 @@
...
+  - OpenCL
+  - OpenMP
+ClangFunctionDifferenceBear:
+  - C
+  - C++
+  - CUDA
+  - Objective-C
+  - Objective-C++
```
Comment (DanPi32) [inline bear-language-requirements.yaml]:
```suggestion
  - Objective-CPP
```

**Comment by sladyn98** [comment]
@jayvdb  Changed the commit message.

**Comment by jayvdb** [comment]
no, you edited the PR page on github.

read https://coala.io/commit and https://coala.io/git , or stackoverflow for amend git commit etc.

**Comment by sladyn98** [comment]
@jayvdb Yes now i used git commit --amend and made the changes.Thanks and kindly review

File: bear-language-requirements.yaml [hunk 27]
```diff
@@ -0,0 +1,409 @@
...
+HAMLLintBear:
+  - Haml
+HTMLHintBear:
+  - HTML
+HTMLLintBear:
+  - HTML
+  - Jinja2
+  - PHP
+HTTPSBear:
+  - true
```
Comment (jayvdb) [inline bear-language-requirements.yaml]: this isnt a language

File: bear-language-requirements.yaml [hunk 28]
```diff
@@ -0,0 +1,409 @@
...
+KeywordBear:
+  - true
+LanguageToolBear:
+  - Natural Language
+LatexLintBear:
+  - Tex
+LicenseCheckBear:
+  - true
+LicenseHeaderBear:
+  - true
```
Comment (jayvdb) [inline bear-language-requirements.yaml]: more ... not a language .. remove

File: .ci/generate_bear_requirements.py [hunk 29]
```diff
@@ -168,6 +171,18 @@ def get_pip_requirements(requirements):
...
+def get_languages(bears):
+    language_dict = {}
+    for bear in bears:
+        language_dict[str(bear.name)] = list(sorted(bear.LANGUAGES))
+    for value in language_dict.values():
+        for n,i in enumerate(value):
+            if(i=='All'):
+                value[n]=True
+        if 'default' in value:
+            value.remove('default')
```
Comment (jayvdb) [inline .ci/generate_bear_requirements.py]: @ayan-b , this is about `DocumentationStyleBear` which uses `DocstyleDefinition.get_available_definitions` - i think you were looking at this .. ?  It was caused by something in coalib ...?

Did you create an issue about it?

c.f. https://github.com/coala/coala-bears/pull/2688#discussion_r225751919

File: .ci/generate_bear_requirements.py [hunk 30]
```diff
@@ -168,6 +171,18 @@ def get_pip_requirements(requirements):
...
     return _get_requirements(requirements, '==')
 
+def get_languages(bears):
+    language_dict = {}
+    for bear in bears:
+        language_dict[str(bear.name)] = list(sorted(bear.LANGUAGES))
+    for value in language_dict.values():
+        for n,i in enumerate(value):
+            if(i=='All'):
+                value[n]=True
```
Comment (jayvdb) [inline .ci/generate_bear_requirements.py]: this is why `true` is going into the yaml .

At https://github.com/coala/coala-bears/pull/2688#discussion_r225752332 , I said that it should be `Bear: true` or `Bear: false`.  It could also be `Bear:`

It can not be
```yaml
Bear:
  - true
```

File: bear-requirements.yaml [hunk 31]
```diff
@@ -231,3 +231,4 @@ r_script_requirements:
     version: '>1.5'
   lintr:
     version: '>=1.0.2'
+
```
Comment (jayvdb) [inline bear-requirements.yaml]: why is this in your diff ????

**Comment by sladyn98** [comment]
@jayvdb I did not make those changes

**Comment by sladyn98** [comment]
@jayvdb how to revert them?

File: bear-requirements.yaml (hunk 31 above)
Comment (li-boxuan) [reply bear-requirements.yaml]: @sladyn98 This is not a newcomer task, so you should be able to use `git ` properly, and people would not give you hints on `git` anymore. I remember in this PR, you add irrelevant stuff multiple times, and every time you say you don't add them 😅 Before you commit or push, verify your local diff is correct.

**Comment by jayvdb** [comment]
It wasnt magic :P  yes, you did it.  and I cant help today - im offline

**Comment by sladyn98** [comment]
Don't you guys think it's because I'm running the beat requirements script
that this file is getting changed?

File: bear-requirements.yaml (hunk 31 above)
Comment (jayvdb) [reply bear-requirements.yaml]: This is still here.

before you push, run `git show` locally.  Read it, top to bottom.  Everything in there is your change.  If it doesnt belong there, do not push.  Fix it.

**Comment by sladyn98** [comment]
Yeah :( I will revert the changes as soon as possible. I am open to
learning as much as I can thanks.

File: .ci/generate_bear_requirements.py [hunk 32]
```diff
@@ -272,6 +297,14 @@ def sort_requirements(req_dict):
     else:
         output = open(args.output, 'w')
 
+    all_bears.sort(key = operator.attrgetter('name'))
```
Comment (jayvdb) [inline .ci/generate_bear_requirements.py]: this doesnt help, and also had whitespace around the `=`

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by sladyn98** [comment]
@jayvdb Okay cool . I learnt a lot from this since this was my first real code contribution and i made a lot of mistakes, but thanks to all the people who helped me @jayvdb @li-boxuan @ishanSrt @Naveenaidu for their input.Looking forward towards learning and contributing more to coala.

**Comment by jayvdb** [comment]
@gitmate-bot ff
